### ISSUE TITLE
When one statement in a multi-statement fails, mysql-otp returns only this one error

### ISSUE THREAD

**Issue by juhlig** [mysql-otp/mysql-otp#169]
If a multi-statement is sent in a `query` call, and at least one of them fails, `mysql-otp` returns (only) an error tuple, even though some of the other statements may have succeeded. To a user, this may appear as if the multi-statement fails atomically, whereas it does not unless used in a transaction.

For illustration, assume the following multi-statement:
```
INSERT INTO example SET foo="bar";
INSERT INTO example SET bar="foo"
```
If, for some reason or other, any one of the two `INSERT`s fails, `mysql-otp` will just return the error tuple for that operation, but the respective other may have succeeded.

Moreover, if _both_ statements fail, only the first error will be returned.

**Comment by zuiderkwast** [comment]
What can be done about it? With or without breaking backward compatibility...

**Comment by juhlig** [comment]
Well, no real brilliant idea... The only thing I can think of is to include the (`ok`s and) `error`s in the return value list of a multi statement (for single statements, everything may remain as is, either `ok`, `{error, Reason}`, or `{ok, Cols, Rows}`)

Like if we issue a query like
```erlang
mysql:query(Conn,
    "SELECT 'foo'; "
    "INSERT INTO existing_table SET foo='bar'; "
    "INSERT INTO nonexisting_table SET foo='bar'; "
    "SELECT 'bar'").
```
it could return something like
```erlang
{ok,
    {[<<"foo">>], [[<<"foo">>]]},
    ok,
    {error, Reason},
    {[<<"bar">>], [[<<"bar">>]]}
}
```

The overall `ok` (on the outer tuple) may be misleading then, but that may be justified somehow. The `ok`s for the individual statements are not strictly needed, but in reality it would be difficult to identify what statement an `error` refers to if we don't include them. In any way, this breaks backwards compatibility.

A solution bridging the two, still breaking but to a lesser degree, would be to include `ok`s and `error`s only if there _are_ errors, and tag the outer tuple with a different atom than `ok`. I feel that this is a bit awkward, though.

Other suggestions? :)

**Comment by zuiderkwast** [comment]
These suggestions are not backwards compatible. I think the only thing that is safe to do is to return the first error as before. A note in the docs would be appropriate though, explaining the behaviour. Perhaps we should recommend against doing INSERTs with auto-id in this way.

**Comment by juhlig** [comment]
I know ^^; But your question was...

> What can be done about it? With or without breaking backward compatibility...

I can't think of any solution that would keep backwards compatibility, so I guess there is no choice but to live with it, as you said 😐

**Comment by juhlig** [comment]
... unless our new MariaDB responsible has any brilliant ideas? @Maria-12648430? 😉

**Comment by Maria-12648430** [comment]
@juhlig I have no idea who you're talking about =^^= But seriously, no, no ideas here either. I just skimmed that part of the code, but it looks like that there is really only the choice to either keep it as it is, change it and become backwards incompatible, or put up new functions so you have both.

**Comment by zuiderkwast** [comment]
Right, adding new functions is an option as well. Not sure it's worth it. But you both ignore my suggested solution: Improve docs. Docs are not to be underestimated. :-)

The only point (AFAIK) of using semicolon-separated queries would be to avoid multiple round-trips to the server. Can we achieve that in some other way? Pipelining would be a cool feature. It would allow processes to send multiple queries using the same client connection without waiting for the previous results to arrive. Eredis does this in a pretty cool way.

**Comment by Maria-12648430** [comment]
> But you both ignore my suggested solution: Improve docs. Docs are not to be underestimated. :-)

No, I understood that as implied in the "keep as it is" option ;)

> The only point (AFAIK) of using semicolon-separated queries would be to avoid multiple round-trips to the server. Can we achieve that in some other way? Pipelining would be a cool feature. It would allow processes to send multiple queries using the same client connection without waiting for the previous results to arrive. Eredis does this in a pretty cool way.

Hm, not sure... I have a feeling that there might be some not-so-obvious corner cases here, so some research and/or tryouts are required.

**Comment by zuiderkwast** [comment]
`gen_server:call` is blocking, but if handle_call returns {noreply, State}, the gen_server isn't blocked and can do other things, such as accepting more gen_server calls.

```
App pid 1   App pid 2                mysql       Server
    |           |                    conn           |
    | gen_server:call                  |            |
    |--------------------------------->| send query |
    |           |                      |===========>|
    |           |              {noreply, State}     |
    |           |                      |            |
    |           | gen_server:call      |            |
    |           |--------------------->| send query |
    |           |                      |===========>|
    |           |              {noreply, State}     |
    |           |                      |            |
    |           |                      |    result  |
    |           |                      |<===========|
    |           |     gen_server:reply |            |
    |<---------------------------------|            |
    |           |                      |    result  |
    |           |     gen_server:reply |<===========|
    |           |<---------------------|            |
```

Assuming that responses arrive from the server in the same order as the queries are sent (which would make sense; the second query is buffered by the server's TCP stack until the server is ready to read it) the above works. Potential problems involve the TCP buffers being filled, but that can happen also when sending a single large query.

With such non-blocking clients, a simple round-robin connection pool like [erlpool](https://github.com/silviucpp/erlpool) can be used efficiently.

Of course it needs to be tested properly and it involves some major rewrite. Maybe it's material for 2.0...

**Comment by juhlig** [comment]
Hm, no, we shouldn't do it that way, the send query as soon as it is issued to mysql_conn part, that is. More to the point, we _can't_ do it that way because we support LOAD LOCAL:

* App Pid 1 issues long-running query, which is immediately sent to the server and executed
* while this is running, App Pid 2 issues a LOAD LOCAL INFILE query, which is immediately sent to the server but not yet read/executed because App Pid 1's query is still in progress
* App Pid 3 issues an arbitrary query which is immediately sent to the server but won't be read/executed since App Pid 2's query is before it
* App Pid 1's query finishes, the server starts App Pid 2's LOAD LOCAL INFILE query, for which it sends a LOAD LOCAL DATA packet to the client
* Instead of a response to the LOAD LOCAL DATA packet, it gets App Pid 3's query packet

And even without that, we shouldn't do it this way:
* App Pid 1 issues long-running query, which is immediately sent to the server and executed
* App Pid 2 issues another long-running query (equally bad, one that produces a big resultset) and crashes
* App Pid 1's query finishes, and App Pid 2's query gets executed even though there is nobody waiting for it any more

Instead, we could batch the queries at the client side, in mysql_conn, and issue them only after we received an ok packet (for inserts/updates/etc) or the header of a resultset.

**Comment by zuiderkwast** [comment]
Good points @juhlig. I wonder if issuing the next query after receiving the header of a result set is much of a gain though. I'm doubt it would be worth the added complexity.

I suppose there are a better ways to avoid the roundtrip latencies resulting from performing too many small queries:

* views
* stored procedures
* semicolon-separated multi-queries
* using joins to fetch more stuff in a single query
